### PR TITLE
Add explicit GUID support for delegates and non-synthesized interfaces.

### LINKED ISSUE #1790: [Authoring] Allow setting GuidAttributes for types
# Proposal: Allow setting `GuidAttributes` for types

## Summary
Currently CsWinRT ignores the `GuidAttribute` set on types, and forcibly uses a generated GUID. This feature request aims to allow the user to specify their own GUID for types.

## Rationale

* This allows explicit GUIDs, so that it is visible what GUID is used in the source code, rather than having to decompile or inspect the metadata of the generated WinMD.
* Match IDL feature sets

## Important Notes
Usage example:
```cs
using System.Runtime.InteropServices;

// The interface below will have a GUID of 123e4567-e89b-12d3-a456-426614174000 in WinMD.

[Guid(“123e4567-e89b-12d3-a456-426614174000”)]
public interface ISample
{ }
```

## Open Questions
N/A

### PR DESCRIPTION
**Pull request by dongle-the-gadget**
I'm not too sure about synthesized interfaces, as it looks like multiple might be generated?

Closes #1790

### CODE REVIEW

**Review by Sergio0694 — COMMENTED** [review]
Can you please also add a couple tests for various cases in the authoring test project?

File: src/Authoring/WinRT.SourceGenerator/WinRTTypeWriter.cs [hunk 1]
```diff
@@ -1620,6 +1613,33 @@ private void AddGuidAttribute(EntityHandle parentHandle, string name)
             AddCustomAttributes("Windows.Foundation.Metadata.GuidAttribute", types, arguments, parentHandle);
         }
 
+        private void AddGuidAttribute(EntityHandle parentHandle, string name)
+        {
+            Guid guid;
+            using (SHA1 sha = new SHA1CryptoServiceProvider())
```
[the remark this answers is not in the document]
Comment (Sergio0694) [reply src/Authoring/WinRT.SourceGenerator/WinRTTypeWriter.cs:1619]: This is a source generator, so it's locked at .NET Standard 2.0, we can't use that API unfortunately.

**Comment by dongle-the-gadget** [comment]
I've added a test in, though I haven't personally tested it myself as I couldn't get CsWinRT to build.

**Comment by Sergio0694** [comment]
CI is not running because this is a fork. I cherry-picked to a branch here to kick it off, it's [here](https://microsoft.visualstudio.com/Dart/_build/results?buildId=107090042&view=results).

**Comment by Sergio0694** [comment]
Fixed a typo, new CI run [here](https://microsoft.visualstudio.com/Dart/_build/results?buildId=107091953&view=results).

**Review by Sergio0694 — APPROVED** [review]
CI is green (just failing on publish, unrelated), changes look good to me.
Thank you! 🙂

File: src/Tests/AuthoringTest/Program.cs [hunk 2]
```diff
@@ -1854,6 +1854,17 @@ private static int GetNumberFromAbi(void* thisPtr, int* value)
...
         }
     }
+
+    [System.Runtime.InteropServices.Guid("26D8EE57-8B1B-46F4-A4F9-8C6DEEEAF53A")]
+    public interface ICustomInterfaceGuid
+    {
+        string HelloWorld();
+    }
+
+    public sealed class CustomInterfaceGuidClass : ICustomInterfaceGuid
```
Comment (manodasanW) [inline src/Tests/AuthoringTest/Program.cs:1864]: Can you add this class to [here](https://github.com/microsoft/CsWinRT/blob/user/sergiopedri/custom-guid/src/Tests/AuthoringConsumptionTest/AuthoringConsumptionTest.exe.manifest).

Comment (dongle-the-gadget) [reply src/Tests/AuthoringTest/Program.cs:1864]: Done

**Review by manodasanW — APPROVED** [review]

**Pull request merged**